subprojects/core/src/integTest/groovy/org/gradle/caching/internal/packaging/BuildCacheEntryPackingIntegrationTest.groovy

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.caching.internal.packaging
+
+import org.gradle.integtests.fixtures.DirectoryBuildCacheFixture
+import org.gradle.integtests.fixtures.daemon.DaemonIntegrationSpec
+import spock.lang.Issue
+import spock.lang.Unroll
+
+class BuildCacheEntryPackingIntegrationTest extends DaemonIntegrationSpec implements DirectoryBuildCacheFixture {
+
+    private static final NON_ASCII_NAME = [
+        "ascii-only": "ascii",
+        "space": " ",
+        "zwnj": "\u200c",
+        "chinese": "敏捷的棕色狐狸跳过了懒狗",
+        "cyrillic": "здравствуйте",
+        "hungarian": "Árvíztűrő tükörfúrógép",
+    ].values().join("-")
+
+    private static final DEFAULT_ENCODINGS = [
+        "UTF-8",
+        "ISO-8859-1",
+        "windows-1250",
+    ]
+
+    @Unroll
+    def "can store and load files having non-ascii characters in file name using default file encoding #fileEncoding"() {
+        def fileName = NON_ASCII_NAME + ".txt"
+        def outputFile = file("dir", fileName)
+
+        buildFile << """
+            println "> Default charset: \${java.nio.charset.Charset.defaultCharset()}"
+            println "> Storing file in cache: $fileName"
+
+            task createFile {
+                outputs.dir("dir")
+                outputs.cacheIf { true }
+                doLast {
+                    file("dir/$fileName").text = "output"
+                }
+            }
+        """
+
+        when:
+        withBuildCache().run("createFile", "-Dfile.encoding=$fileEncoding")
+        then:
+        output.contains("> Default charset: $fileEncoding")
+        executedAndNotSkipped(":createFile")
+
+        when:
+        assert outputFile.delete()
+        withBuildCache().run("createFile", "-Dfile.encoding=$fileEncoding")
+        skipped(":createFile")
+
+        then:
+        output.contains("> Default charset: $fileEncoding")
+        outputFile.text == "output"
+
+        where:
+        fileEncoding << DEFAULT_ENCODINGS
+    }
+
+    @Issue("https://github.com/gradle/gradle/issues/9877")
+    @Unroll
```

Comment:
❓ Would it make sense to add the `@Unroll` to the class instead?
❓ Should we add the `Issue` annotation to the rest of the related tests?

Reply:
I moved the `@Issue` annotation to where it belongs. I think `@Unroll` on the class means something a bit different, so I don't want to move it there. We might add more tests here that don't need to be unrolled.

Reply:
Do I understand it correctly that we have tests with `where` clauses that we don't want to unroll?

Reply:
No, there are multiple reasons why I don't usually put `@Unroll` on a class.

IntelliJ has a problem navigating to unrolled tests. Adding `@Unroll` to the test class used to mean that navigation to non-parametrized tests was broken, too. This latter problem seems to have been fixed lately, so we _could_ add `@Unroll` to this class as you suggest.

But then again, why add it to the test class? If I move all parametrized tests to a different class, should I then remove `@Unroll` from the class itself? It's not the right abstraction level to annotate. What we need instead is global-unroll that enables unrolling automatically for every test. There seems to be a Spock extension to do just that: https://github.com/szpak/spock-global-unroll, let's investigate that outside of this PR.

Reply:
Ah, that's brilliant, lemme' see how complicated that would be

Reply:
Not worth time in it since: https://github.com/spockframework/spock/issues/967#issuecomment-497016452